src/contracts/libraries/BN254.sol

```diff
@@ -127,6 +131,49 @@ library BN254 {
         require(success, "ec-add-failed");
     }
 
+    /**
+     * @notice an optimized ecMul implementation that takes O(log_2(s)) ecAdds
+     * @param p the point to multiply
+     * @param s the scalar to multiply by
+     * @dev this function is only safe to use if the scalar is 9 bits or less
+     */ 
+    function scalar_mul_tiny(BN254.G1Point memory p, uint16 s) internal view returns (BN254.G1Point memory) {
+        require(s < 2**9, "scalar-too-large");
+
+        // if s is 1 return p
+        if(s == 1) {
+            return p;
+        }
+
+        // the accumulated product to return
+        BN254.G1Point memory acc = BN254.G1Point(0, 0);
+        // the 2^n*p to add to the accumulated product in each iteration
+        BN254.G1Point memory p2n = p;
+        // value of most signifigant bit
+        uint16 m = 1;
+        // index of most signifigant bit
+        uint8 i = 0;
+
+        //loop until we reach the most signifigant bit
+        while(s > m){
+            unchecked {
+                // if the  current bit is 1, add the 2^n*p to the accumulated product
+                if (s >> i & 1 == 1) {
+                    acc = plus(acc, p2n);
+                }
+                // double the 2^n*p for the next iteration
+                p2n = plus(p2n, p2n);
+
+                // increment the index and double the value of the most signifigant bit
+                m <<= 1;
+                ++i;
```

Comment:
so `m` is always equal to `1 << i`, right?
seems like a potential simplification, presumably  just need to make sure i is a larger integer type to avoid overflowing the uint8 type when left-shifting.